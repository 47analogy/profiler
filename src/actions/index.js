export function waitingForProfileFromAddon() {
  return {
    type: 'WAITING_FOR_PROFILE_FROM_ADDON'
  };
}

export function receiveProfileFromAddon(profile) {
  return {
    type: 'RECEIVE_PROFILE_FROM_ADDON',
    profile: profile
  };
}

export function requestingSymbolTable(requestedLib) {
  return {
    type: 'REQUESTING_SYMBOL_TABLE',
    requestedLib
  };
}

export function receivedSymbolTableReply(requestedLib) {
  return {
    type: 'RECEIVED_SYMBOL_TABLE_REPLY',
    requestedLib
  };
}

export function startSymbolicating() {
  return {
    type: 'START_SYMBOLICATING'
  };
}

export function profileSymbolicationStep(profile) {
  return {
    type: 'PROFILE_SYMBOLICATION_STEP',
    profile
  };
}

export function doneSymbolicating() {
  return { type: 'DONE_SYMBOLICATING' };
}

export function mergeFunctions(threadIndex, oldFuncToNewFuncMap) {
  return {
    type: 'MERGE_FUNCTIONS',
    threadIndex, oldFuncToNewFuncMap
  };
}

export function assignFunctionNames(threadIndex, funcIndices, funcNames) {
  return {
    type: 'ASSIGN_FUNCTION_NAMES',
    threadIndex, funcIndices, funcNames
  };
}

export function changeSelectedFuncStack(selectedFuncStack) {
  return {
    type: 'CHANGE_SELECTED_FUNC_STACK',
    selectedFuncStack
  };
}
