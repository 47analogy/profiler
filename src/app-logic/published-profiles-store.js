/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
// @flow

// This file contains the code responsible for storing informations about
// published profiles.

import { openDB, deleteDB } from 'idb';
import type { DB as Database } from 'idb';
import type { StartEndRange } from 'firefox-profiler/types';

export type ProfileData = {|
  +profileToken: string, // This is the primary key.
  +jwtToken: string | null,
  +publishedDate: Date,
  +name: string,
  +preset: string | null,
  +originHostname: string | null,
  +meta: {|
    // We're using some of the properties of the profile meta, but we're not
    // reusing the type ProfileMeta completely because we don't want to be
    // impacted from future changes to ProfileMeta.
    // Look at ProfileMeta definition to know more about these fields.
    +product: string,
    +abi?: string,
    +platform?:
      | 'Android'
      | 'Windows'
      | 'Macintosh'
      // X11 is used for historic reasons, but this value means that it is a Unix platform.
      | 'X11'
      | string,
    +toolkit?: string,
    +misc?: string,
    +oscpu?: string,
    +toolkit?: 'gtk' | 'windows' | 'cocoa' | 'android' | string,
    +updateChannel?:
      | 'default' // Local builds
      | 'nightly'
      | 'nightly-try' // Nightly try builds for QA
      | 'aurora' // Developer Edition channel
      | 'beta'
      | 'release'
      | 'esr' // Extended Support Release channel
      | string,
    +appBuildID?: string,
  |},
  // Storing the state as the path makes it easy to reuse our URL upgrade mechanism.
  +urlPath: string,
  +publishedRange: StartEndRange,
|};

// Exported for tests.
export const DATABASE_NAME = 'published-profiles-store';
export const OBJECTSTORE_NAME = 'published-profiles';
export const DATABASE_VERSION = 1;

async function reallyOpen(): Promise<Database> {
  const db = await openDB(DATABASE_NAME, DATABASE_VERSION, {
    upgrade(db) {
      const store = db.createObjectStore(OBJECTSTORE_NAME, {
        keyPath: 'profileToken',
      });
      store.createIndex('originHostname', 'originHostname');
    },
    blocked() {
      // Ideally we should throw this error, but idb doesn't allow this, so
      // instead we just output this in the console.
      // See https://github.com/jakearchibald/idb/issues/186
      console.error(
        new Error(
          'The published profiles store database could not be upgraded because it is ' +
            'open in another tab. Please close all your other profiler.firefox.com ' +
            'tabs and refresh.'
        )
      );
    },
  });

  return db;
}

async function open(): Promise<Database> {
  if (!window.indexedDB) {
    throw new Error('Could not find indexedDB on the window object.');
  }

  let db;
  try {
    db = await reallyOpen();
  } catch (e) {
    if (e.name === 'VersionError') {
      // This error fires if the database already exists, and the existing
      // database has a higher version than what we requested. So either
      // this version of profiler.firefox.com is outdated, or somebody briefly tried
      // to change this database format (and increased the version number)
      // and then downgraded to a version of profiler.firefox.com without those
      // changes.
      // We delete the database and try again.
      await deleteDB(DATABASE_NAME);
      db = await reallyOpen();
    } else {
      throw e;
    }
  }

  return db;
}

export async function storeProfileData(
  profileData: ProfileData
): Promise<void> {
  const db = await open();
  await db.put(OBJECTSTORE_NAME, profileData);
}

export async function listAllProfileData(): Promise<ProfileData[]> {
  const db = await open();
  return db.getAll(OBJECTSTORE_NAME);
}

export async function retrieveProfileData(
  profileToken: string
): Promise<ProfileData> {
  const db = await open();
  return db.get(OBJECTSTORE_NAME, profileToken);
}
