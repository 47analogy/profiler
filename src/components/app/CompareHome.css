.compareHome {
  width: 75%;
  margin: auto;
  padding: 8em;

  border: 1px solid #CCC;
  border-radius: 3px;
  box-shadow: 0 5px 25px #0b1f50;
  background: #fff;
}

.compareHomeForm {
  display: grid;
  grid-template-columns: auto 1fr;
  grid-gap: 1em;
  align-items: center;
}

.compareHomeSubmitButton {
  grid-column-start: span 2
}

.compareHomeFormLabel {
  white-space: nowrap;
}

@media (max-width: 450px) {
  .compareHome {
    width: auto;
    padding: 1em;
  }
}
