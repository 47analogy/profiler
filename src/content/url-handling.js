import queryString from 'query-string';
import { stringifyRangeFilters, parseRangeFilters } from './range-filters';
import { stringifyCallTreeFilters, parseCallTreeFilters } from './call-tree-filters';

// {
//   // general:
//   dataSource: 'from-addon', 'local', 'public',
//   hash: '' or 'aoeurschsaoeuch',
//   selectedTab: 'summary' or 'calltree' or ...,
//   rangeFilters: [] or [{ start, end }, ...],
//   selectedThread: 0 or 1 or ...,
//   
//   // only when selectedTab === 'calltree':
//   callTreeSearchString: '' or '::RunScript' or ...,
//   callTreeFilters: [[], [{type:'prefix', matchJSOnly:true, prefixFuncs:[1,3,7]}, {}, ...], ...], // one per thread
//   jsOnly: false or true,
//   invertCallstack: false or true,
// }

function dataSourceDirs(urlState) {
  const { dataSource } = urlState;
  switch (dataSource) {
    case 'from-addon':
      return ['from-addon'];
    case 'local':
      return ['local', urlState.hash];
    case 'public':
      return ['public', urlState.hash];
    default:
      return [];
  }
}

export function urlFromState(urlState) {
  const { dataSource } = urlState;
  if (dataSource === 'none') {
    return '/';
  }
  const pathname = '/' + [
    ...dataSourceDirs(urlState),
    urlState.selectedTab,
  ].join('/') + '/';
  const query = Object.assign({
    range: stringifyRangeFilters(urlState.rangeFilters) || undefined,
    thread: `${urlState.selectedThread}`,
  }, urlState.selectedTab === 'calltree' ? {
    search: urlState.callTreeSearchString || undefined,
    invertCallstack: urlState.invertCallstack ? null : undefined,
    jsOnly: urlState.jsOnly ? null : undefined,
    callTreeFilters: stringifyCallTreeFilters(urlState.callTreeFilters[urlState.selectedThread]) || undefined,
  } : {});
  const qString = queryString.stringify(query);
  return pathname + (qString ? '?' + qString : '');
}

export function stateFromURL(url) {
  const [pathname, qString] = url.split('?');
  const dirs = pathname.split('/').filter(d => d);
  const dataSource = dirs[0] || 'none';
  if (!['none', 'from-addon', 'local', 'public'].includes(dataSource)) {
    throw new Error('unexpected data source');
  }
  const needHash = ['local', 'public'].includes(dataSource);
  const query = queryString.parse(qString);
  const selectedThread = query.thread !== undefined ? +query.thread : 0;
  return {
    dataSource,
    hash: needHash ? dirs[1] : '',
    selectedTab: (needHash ? dirs[2] : dirs[1]) || 'calltree',
    rangeFilters: query.range ? parseRangeFilters(query.range) : [],
    selectedThread: selectedThread,
    callTreeSearchString: query.search || '',
    callTreeFilters: {
      [selectedThread]: query.callTreeFilters ? parseCallTreeFilters(query.callTreeFilters) : [],
    },
    jsOnly: query.jsOnly !== undefined,
    invertCallstack: query.invertCallstack !== undefined,
  };
}
