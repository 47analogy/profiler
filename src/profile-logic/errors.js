/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import type { Library } from './symbol-store';

// Used during the symbolication process to express that we couldn't find
// symbols for a specific library
export class SymbolsNotFoundError extends Error {
  library: Library;
  error: ?Error;

  constructor(message: string, library: Library, error?: Error) {
    super(message);
    this.name = 'SymbolsNotFoundError';
    this.library = library;
    this.error = error;
  }
}
