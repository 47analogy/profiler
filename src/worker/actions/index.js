import { summarizeProfile } from '../../common/summarize-profile';

export function processProfileSummary() {
  return function (dispatch, getState) {
    dispatch({
      toContent: true,
      type: 'PROFILE_SUMMARY_PROCESSED',
      summary: summarizeProfile(getState().profile),
    });
  };
}

export function profileProcessed(profile) {
  return {
    type: 'PROFILE_PROCESSED',
    profile: profile,
  };
}
