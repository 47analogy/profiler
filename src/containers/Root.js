import React, { Component, PropTypes } from 'react';
import { connect, Provider } from 'react-redux';
import { Router, Route, Redirect, IndexRoute } from 'react-router';
import * as actions from '../actions';
import ProfileViewer from '../components/ProfileViewer';
import Home from '../containers/Home';

class GettingProfileFromAddonImpl extends Component {
  componentDidMount() {
    this.props.retrieveProfileFromAddon(this.props.location);
  }

  render() {
    if (window.geckoProfilerPromise) {
      return (
        <div>Retrieving profile from the gecko profiler addon...</div>
      );
    }
    return <p className='profilerNotConnectedErrorMessage'>This tab is not connected to the gecko profiler addon.</p>;
  }
}

GettingProfileFromAddonImpl.propTypes = {
  retrieveProfileFromAddon: PropTypes.func.isRequired,
  location: PropTypes.object.isRequired,
};

const GettingProfileFromAddon = connect(() => ({}), actions)(GettingProfileFromAddonImpl);

const GettingProfileFromLocal = () => (<div>Not implemented yet.</div>);

class GettingProfileFromWebImpl extends Component {
  componentDidMount() {
    const { location, params, retrieveProfileFromWeb } = this.props;
    if (params.hash) {
      retrieveProfileFromWeb(params.hash, location);
    }
  }

  render() {
    const { location, params } = this.props;
    if (params.hash) {
      return (
        <div>Retrieving profile from the public profile store...</div>
      );
    }
    return <FourOhFour location={location}/>;
  }
}

GettingProfileFromWebImpl.propTypes = {
  location: PropTypes.object.isRequired,
  params: PropTypes.object.isRequired,
  retrieveProfileFromWeb: PropTypes.func.isRequired,
};

const GettingProfileFromWeb = connect(() => ({}), actions)(GettingProfileFromWebImpl);

const FourOhFour = ({ location }) => {
  return (
    <div>There is no route handler for the URL {location.pathname}</div>
  );
};

FourOhFour.propTypes = {
  location: PropTypes.object.isRequired,
};

const ProfileViewerWithDataFromAddon = connect(({ view }) => ({ view }))(({ view, location, params }) => {
  switch (view) {
    case 'INITIALIZING':
      return <GettingProfileFromAddon params={params} location={location} />;
    case 'PROFILE':
      return <ProfileViewer dataSource='from-addon' params={params} location={location}/>;
    default:
      return <div>View not found.</div>;
  }
});

const ProfileViewerWithDataFromLocal = connect(({ view }) => ({ view }))(({ view, location, params }) => {
  switch (view) {
    case 'INITIALIZING':
      return <GettingProfileFromLocal params={params} location={location} />;
    case 'PROFILE':
      return <ProfileViewer dataSource='local' params={params} location={location}/>;
    default:
      return <div>View not found.</div>;
  }
});

const ProfileViewerWithDataFromWeb = connect(({ view }) => ({ view }))(({ view, location, params }) => {
  switch (view) {
    case 'INITIALIZING':
      return <GettingProfileFromWeb params={params} location={location} />;
    case 'PROFILE':
      return <ProfileViewer dataSource='public' params={params} location={location}/>;
    default:
      return <div>View not found.</div>;
  }
});

const PassThrough = ({ children }) => children;

export default class Root extends Component {
  render() {
    const { store, history } = this.props;
    return (
      <Provider store={store}>
        <Router history={history}>
          <Route path='/' component={PassThrough}>
            <IndexRoute component={Home} />
            <Route path='from-addon/:selectedTab' component={ProfileViewerWithDataFromAddon} />
            <Redirect from='from-addon' to='/from-addon/calltree/' component={ProfileViewerWithDataFromAddon} />
            <Route path='local/:hash/:selectedTab' component={ProfileViewerWithDataFromLocal} />
            <Redirect from='local/:hash' to='/local/:hash/calltree/' component={ProfileViewerWithDataFromLocal} />
            <Route path='public/:hash/:selectedTab' component={ProfileViewerWithDataFromWeb} />
            <Redirect from='public/:hash' to='/public/:hash/calltree/' component={ProfileViewerWithDataFromWeb} />
            <Route path='*' component={FourOhFour} />
          </Route>
        </Router>
      </Provider>
    );
  }
}

Root.propTypes = {
  store: PropTypes.any.isRequired,
  history: PropTypes.any.isRequired,
};
