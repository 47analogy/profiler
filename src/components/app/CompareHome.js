/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import React, { PureComponent } from 'react';
import { changeProfilesToCompare } from '../../actions/app';

import explicitConnect from '../../utils/connect';
import type {
  ExplicitConnectOptions,
  ConnectedProps,
} from '../../utils/connect';

import './CompareHome.css';

type DispatchProps = {|
  +changeProfilesToCompare: typeof changeProfilesToCompare,
|};

type Props = ConnectedProps<{||}, {||}, DispatchProps>;

type State = {
  profile1: string,
  profile2: string,
};

class CompareHome extends PureComponent<Props, State> {
  state = { profile1: '', profile2: '' };

  handleInputChange = (event: SyntheticInputEvent<>) => {
    const { name, value } = event.target;
    this.setState({ [name]: value });
  };

  handleFormSubmit = (e: SyntheticEvent<>) => {
    e.preventDefault();
    let { profile1, profile2 } = this.state;
    const { changeProfilesToCompare } = this.props;
    if (!profile1.startsWith('http')) {
      profile1 = 'http://' + profile1;
    }
    if (!profile2.startsWith('http')) {
      profile2 = 'http://' + profile2;
    }
    changeProfilesToCompare(profile1, profile2);
  };

  render() {
    const { profile1, profile2 } = this.state;

    return (
      <form className="compareHome" onSubmit={this.handleFormSubmit}>
        <h1>Enter the 2 profile URLs that youâ€™d like to compare</h1>
        <p>
          The tool will extract the data from the selected track and range for
          each profile, and put them both on the same view to make them easy to
          compare.
        </p>
        <label htmlFor="compareHomeProfile1">Profile 1:</label>
        <input
          name="profile1"
          id="compareHomeProfile1"
          className="compareHome-input"
          onChange={this.handleInputChange}
          value={profile1}
        />
        <label htmlFor="compareHomeProfile2">Profile 2:</label>
        <input
          name="profile2"
          id="compareHomeProfile2"
          className="compareHome-input"
          onChange={this.handleInputChange}
          value={profile2}
        />
        <input
          className="compareHome-submit-button"
          type="submit"
          value="Retrieve profiles"
        />
      </form>
    );
  }
}

const options: ExplicitConnectOptions<{||}, {||}, Props> = {
  mapDispatchToProps: { changeProfilesToCompare },
  component: CompareHome,
};

export default explicitConnect(options);
