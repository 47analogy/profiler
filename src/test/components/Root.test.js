/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import * as React from 'react';

import { ProfileViewWhenReady } from '../../components/app/Root';
import { updateUrlState } from '../../actions/app';
import {
  fatalError,
  temporaryError,
  viewProfile,
  waitingForProfileFromAddon,
  waitingForProfileFromStore,
} from '../../actions/receive-profile';
import { stateFromLocation } from '../../app-logic/url-handling';
import { TemporaryError } from '../../utils/errors';

import { shallowWithStore } from '../fixtures/enzyme';
import { blankStore } from '../fixtures/stores';
import { getProfileFromTextSamples } from '../fixtures/profiles/make-profile';

describe('app/ProfileViewWhenReady', function() {
  it('renders an initial home', function() {
    const { view } = setup();

    // dive() will shallow-render the wrapped component
    expect(view.dive()).toMatchSnapshot();
  });

  it('renders the addon loading page', function() {
    const { view, dispatch, navigateToAddonLoadingPage } = setup();

    navigateToAddonLoadingPage();
    dispatch(waitingForProfileFromAddon());
    expect(view.dive()).toMatchSnapshot();
  });

  it('renders the profile view', function() {
    const { view, dispatch, navigateToStoreLoadingPage } = setup();

    navigateToStoreLoadingPage();
    dispatch(waitingForProfileFromStore());
    expect(view.dive()).toMatchSnapshot();

    const { profile } = getProfileFromTextSamples(`A`);
    dispatch(viewProfile(profile));
    expect(view.dive()).toMatchSnapshot();
  });

  it('renders temporary errors', function() {
    const { view, dispatch, navigateToStoreLoadingPage } = setup();

    navigateToStoreLoadingPage();
    dispatch(
      temporaryError(
        new TemporaryError('This is a temporary error, wait a bit more.', {
          count: 3,
          total: 10,
        })
      )
    );
    expect(view.dive()).toMatchSnapshot();
  });

  it('renders an home when the user pressed back after an error', function() {
    const {
      view,
      dispatch,
      navigateToStoreLoadingPage,
      navigateBackToHome,
    } = setup();

    navigateToStoreLoadingPage();
    dispatch(fatalError(new Error('Error while loading profile')));
    expect(view.dive()).toMatchSnapshot();
    expect(console.error).toHaveBeenCalled();

    navigateBackToHome();
    expect(view.dive()).toMatchSnapshot();
  });
});

function setup() {
  // Let's silence the error output to the console
  jest.spyOn(console, 'error').mockImplementation(() => {});

  const store = blankStore();
  const view = shallowWithStore(<ProfileViewWhenReady />, store);

  function navigateToStoreLoadingPage() {
    const newUrlState = stateFromLocation({
      pathname: '/public/404ff08ee5f412f355264601f12e5feff0ebaee6/calltree',
      search: '',
      hash: '',
    });
    store.dispatch(updateUrlState(newUrlState));
  }

  function navigateToAddonLoadingPage() {
    const newUrlState = stateFromLocation({
      pathname: '/from-addon/',
      search: '',
      hash: '',
    });
    store.dispatch(updateUrlState(newUrlState));
  }

  function navigateBackToHome() {
    const newUrlState = stateFromLocation({
      pathname: '/',
      hash: '',
      search: '',
    });
    store.dispatch(updateUrlState(newUrlState));
  }

  return {
    view,
    dispatch: store.dispatch,
    navigateToStoreLoadingPage,
    navigateToAddonLoadingPage,
    navigateBackToHome,
  };
}
