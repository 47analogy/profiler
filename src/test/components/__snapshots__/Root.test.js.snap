// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app/ProfileViewWhenReady does not try to retrieve a profile when moving from from-addon to public 1`] = `<profile-viewer />`;

exports[`app/ProfileViewWhenReady renders a compare home when navigating to /compare/, then loads when changing profiles 1`] = `
<main
  class="compareHome"
>
  <h1
    class="compareHomeTitle"
  >
    Enter the profile URLs that youâ€™d like to compare
  </h1>
  <p>
    The tool will extract the data from the selected track and range for each profile, and put them both on the same view to make them easy to compare.
  </p>
  <form
    class="compareHomeForm"
  >
    <label
      class="compareHomeFormLabel"
      for="compareHomeProfile1"
    >
      Profile 1:
    </label>
    <input
      class="photon-input"
      id="compareHomeProfile1"
      name="profile1"
      placeholder="http://"
      required=""
      type="url"
      value=""
    />
    <label
      class="compareHomeFormLabel"
      for="compareHomeProfile2"
    >
      Profile 2:
    </label>
    <input
      class="photon-input"
      id="compareHomeProfile2"
      name="profile2"
      placeholder="http://"
      required=""
      type="url"
      value=""
    />
    <input
      class="compareHomeSubmitButton"
      type="submit"
      value="Retrieve profiles"
    />
  </form>
</main>
`;

exports[`app/ProfileViewWhenReady renders a compare home when navigating to /compare/, then loads when changing profiles 2`] = `
<div
  class="rootMessageContainer"
>
  <div
    class="rootMessage"
  >
    <h1
      class="rootMessageTitle"
    >
      perf.html
    </h1>
    <div
      class="rootMessageText"
    >
      Reading and processing profiles...
    </div>
    <div
      class="loading"
    >
      <div
        class="loading-div loading-div-1 loading-row-1"
      />
      <div
        class="loading-div loading-div-2 loading-row-2"
      />
      <div
        class="loading-div loading-div-3 loading-row-3"
      />
      <div
        class="loading-div loading-div-4 loading-row-3"
      />
      <div
        class="loading-div loading-div-5 loading-row-4"
      />
      <div
        class="loading-div loading-div-6 loading-row-4"
      />
      <div
        class="loading-div loading-div-7 loading-row-4"
      />
      <div
        class="loading-div loading-div-8 loading-row-4"
      />
      <div
        class="loading-div loading-div-9 loading-row-4"
      />
      <div
        class="loading-div loading-div-10 loading-row-4"
      />
    </div>
  </div>
</div>
`;

exports[`app/ProfileViewWhenReady renders an home when the user pressed back after an error 1`] = `
<div
  class="rootMessageContainer"
>
  <div
    class="rootMessage"
  >
    <h1
      class="rootMessageTitle"
    >
      perf.html
    </h1>
    <div
      class="rootMessageText"
    >
      Could not download the profile.
    </div>
    <div
      class="rootMessageAdditional"
    >
      <p>
        Error: Error while loading profile
      </p>
      <p>
        The full stack has been written to the Web Console.
      </p>
      <a
        href="/"
      >
        Back to home
      </a>
    </div>
  </div>
</div>
`;

exports[`app/ProfileViewWhenReady renders an home when the user pressed back after an error 2`] = `<home />`;

exports[`app/ProfileViewWhenReady renders an initial home 1`] = `<home />`;

exports[`app/ProfileViewWhenReady renders temporary errors 1`] = `
<div
  class="rootMessageContainer"
>
  <div
    class="rootMessage"
  >
    <h1
      class="rootMessageTitle"
    >
      perf.html
    </h1>
    <div
      class="rootMessageText"
    >
      Downloading and processing the profile...
    </div>
    <div
      class="rootMessageAdditional"
    >
      <p>
        This is a temporary error, wait a bit more.
      </p>
      <p>
        Tried 3 times out of 10.
      </p>
      <a
        href="/"
      >
        Back to home
      </a>
    </div>
    <div
      class="loading"
    >
      <div
        class="loading-div loading-div-1 loading-row-1"
      />
      <div
        class="loading-div loading-div-2 loading-row-2"
      />
      <div
        class="loading-div loading-div-3 loading-row-3"
      />
      <div
        class="loading-div loading-div-4 loading-row-3"
      />
      <div
        class="loading-div loading-div-5 loading-row-4"
      />
      <div
        class="loading-div loading-div-6 loading-row-4"
      />
      <div
        class="loading-div loading-div-7 loading-row-4"
      />
      <div
        class="loading-div loading-div-8 loading-row-4"
      />
      <div
        class="loading-div loading-div-9 loading-row-4"
      />
      <div
        class="loading-div loading-div-10 loading-row-4"
      />
    </div>
  </div>
</div>
`;

exports[`app/ProfileViewWhenReady renders the addon loading page, and the profile view after capturing 1`] = `
<div
  class="rootMessageContainer"
>
  <div
    class="rootMessage"
  >
    <h1
      class="rootMessageTitle"
    >
      perf.html
    </h1>
    <div
      class="rootMessageText"
    >
      Grabbing the profile from the Gecko Profiler Addon...
    </div>
    <div
      class="loading"
    >
      <div
        class="loading-div loading-div-1 loading-row-1"
      />
      <div
        class="loading-div loading-div-2 loading-row-2"
      />
      <div
        class="loading-div loading-div-3 loading-row-3"
      />
      <div
        class="loading-div loading-div-4 loading-row-3"
      />
      <div
        class="loading-div loading-div-5 loading-row-4"
      />
      <div
        class="loading-div loading-div-6 loading-row-4"
      />
      <div
        class="loading-div loading-div-7 loading-row-4"
      />
      <div
        class="loading-div loading-div-8 loading-row-4"
      />
      <div
        class="loading-div loading-div-9 loading-row-4"
      />
      <div
        class="loading-div loading-div-10 loading-row-4"
      />
    </div>
  </div>
</div>
`;

exports[`app/ProfileViewWhenReady renders the addon loading page, and the profile view after capturing 2`] = `<profile-viewer />`;

exports[`app/ProfileViewWhenReady renders the profile view 1`] = `
<div
  class="rootMessageContainer"
>
  <div
    class="rootMessage"
  >
    <h1
      class="rootMessageTitle"
    >
      perf.html
    </h1>
    <div
      class="rootMessageText"
    >
      Downloading and processing the profile...
    </div>
    <div
      class="loading"
    >
      <div
        class="loading-div loading-div-1 loading-row-1"
      />
      <div
        class="loading-div loading-div-2 loading-row-2"
      />
      <div
        class="loading-div loading-div-3 loading-row-3"
      />
      <div
        class="loading-div loading-div-4 loading-row-3"
      />
      <div
        class="loading-div loading-div-5 loading-row-4"
      />
      <div
        class="loading-div loading-div-6 loading-row-4"
      />
      <div
        class="loading-div loading-div-7 loading-row-4"
      />
      <div
        class="loading-div loading-div-8 loading-row-4"
      />
      <div
        class="loading-div loading-div-9 loading-row-4"
      />
      <div
        class="loading-div loading-div-10 loading-row-4"
      />
    </div>
  </div>
</div>
`;

exports[`app/ProfileViewWhenReady renders the profile view 2`] = `<profile-viewer />`;
